docs/primitives.md

```diff
@@ -249,23 +249,26 @@ Slice function interfaces:
    * width: constant slice width
 
 
-### set_bit
+### set_index
```

Comment:
I would be in favor of renaming the function (publicly, i.e. in the docs) to `set_bit`. We could deprecate `set_index` for now with warning (doubt anyone is using it anyway...)

Reply:
It was changed to set_index because the pattern is more general than just set_bit (i.e. you can use it on generic arrays, rather than just on Bits), so I think it makes more sense to keep it as set_index for generality

Reply:
I see. In that case, why do we need a separate function? I.e. why is it not covered by set slice with width 1?

Reply:
Makes sense -- but I would argue `x[1]` and `x[1:2]` is the same interface with different arguments (in fact it literally is the function `__getitem__`).

Maybe we don't need to do this in this change, but seems reasonable to refactor to have the following interface:

```python
"""
* target: Array[N, T]
* value: Union[Array[M, T], T]
* start: UInt[W]

where M < N and W <= clog2(N)
"""
def set(target: Array, value: Union[Type, Array], start: UInt):
    ...
```

Now that I think about it there are few new concerns here:
* Should this function work with all `Array`s and not just `Bits`? I kind of assumed it did, but looking at the code closer I'm not certain.
* Do we need a `width` argument? Can it just be inferred from `value`? If `value` is an array of the same dimensionality as `target` then it is a slice set; else if it is of dimensionality one less than `value` then it is an index set (of course assuming the contained types match up and all).
* In the case `start` being a constant value, should we optimize for the simple wiring (non-dynamic mux) case?